File: src/RoboticusDebugger.cpp
#include "RoboticusDebugger.h"

bool Sensor::isTriggered() const {
  if (operatorStr == ">=")
    return input >= threshold;
  if (operatorStr == "<=")
    return input <= threshold;
  if (operatorStr == ">")
    return input > threshold;
  if (operatorStr == "<")
    return input < threshold;
  if (operatorStr == "==")
    return input == threshold;
  if (operatorStr == "!=")
    return input != threshold;
  return false;
}

RoboticusDebugger::RoboticusDebugger(Print &output)
    : _output(output), _hasData(false) {}

void RoboticusDebugger::begin() {
  _doc.clear();
  _sensorsArray = _doc[0].to<JsonArray>();
  _vectorsArray = _doc[1].to<JsonArray>();
  _hasData = true;
}

void RoboticusDebugger::addSensor(const String &name, float input,
                                  const String &operatorStr, float threshold,
                                  const String &layer, float x, float y) {
  if (!_hasData)
    begin();

  Sensor sensor{name, input, operatorStr, threshold, layer, x, y};
  add(sensor);
}

void RoboticusDebugger::addVector(const String &name, float rotation,
                                  const String &color, const String &layer,
                                  float x, float y) {
  if (!_hasData)
    begin();

  Vector vector{name, rotation, color, layer, x, y};
  add(vector);
}

void RoboticusDebugger::add(const Sensor &sensor) {
  if (!_hasData)
    begin();

  JsonArray obj = _sensorsArray.add<JsonArray>();
  // [0]=name [1]=input [2]=isTriggered [3]=threshold [4]=layer [5]=x [6]=y
  obj.add(sensor.name);
  obj.add(sensor.input);
  obj.add(sensor.isTriggered());
  obj.add(sensor.threshold);
  obj.add(sensor.layer);
  obj.add(sensor.x);
  obj.add(sensor.y);
}

void RoboticusDebugger::add(const Vector &vector) {
  if (!_hasData)
    begin();

  JsonArray obj = _vectorsArray.add<JsonArray>();
  // [0]=name [1]=rotation [2]=color [3]=layer [4]=x [5]=y
  obj.add(vector.name);
  obj.add(vector.rotation);
  obj.add(vector.color);
  obj.add(vector.layer);
  obj.add(vector.x);
  obj.add(vector.y);
}
void RoboticusDebugger::write() {
  if (!_hasData)
    return;

  _doc[2] = millis();
  _doc.shrinkToFit();

  size_t msgSize = measureMsgPack(_doc);

  uint8_t header[3];
  header[0] = 0xFD;
  header[1] = (uint8_t)(msgSize & 0xFF);
  header[2] = (uint8_t)((msgSize >> 8) & 0xFF);

  // unsigned long start = micros(); // Start timer
  _output.write(header, 3);
  serializeMsgPack(_doc, _output);
  // unsigned long end = micros(); // End timer

  // Serial.print("Time (us): ");
  // Serial.println(end - start);
  // Serial.print("Bytes: ");
  // Serial.println(msgSize);

  _hasData = false;
}

bool RoboticusDebugger::isEmpty() const { return !_hasData; }

File: python/RoboticusDebugger.py
import sys
import time
import struct
import msgpack # pip install msgpack
 
_START_TIME = time.monotonic()
 
class Sensor:
    def __init__(self, name: str, input_val: float, operator_str: str,
                 threshold: float, layer: str, x: float, y: float):
        self.name = name
        self.input = float(input_val)
        self.operator_str = operator_str
        self.threshold = float(threshold)
        self.layer = layer
        self.x = float(x)
        self.y = float(y)
 
    def is_triggered(self) -> bool:
        ops = {
            ">=": self.input >= self.threshold,
            "<=": self.input <= self.threshold,
            ">":  self.input >  self.threshold,
            "<":  self.input <  self.threshold,
            "==": self.input == self.threshold,
            "!=": self.input != self.threshold,
        }
        return ops.get(self.operator_str, False)
 
 
class Vector:
    def __init__(self, name: str, rotation: float, color: str,
                 layer: str, x: float, y: float):
        self.name = name
        self.rotation = float(rotation)
        self.color = color
        self.layer = layer
        self.x = float(x)
        self.y = float(y)
 
 
class RoboticusDebugger:
    def __init__(self, output=None):
        self._output = output
        self._sensors: list = []
        self._vectors: list = []
        self._has_data = False
 
    def add(self, item):
        if isinstance(item, Sensor):
            self._add_sensor_to_pack(item)
        elif isinstance(item, Vector):
            self._add_vector_to_pack(item)
        else:
            raise TypeError(f"Expected Sensor or Vector, got {type(item)}")
 
    def add_sensor(self, name, input_val, operator_str, threshold, layer, x, y):
        self.add(Sensor(name, input_val, operator_str, threshold, layer, x, y))
 
    def add_vector(self, name, rotation, color, layer, x, y):
        self.add(Vector(name, rotation, color, layer, x, y))
 
    def write(self):
        if not self._has_data:
            return
 
        frame = [self._sensors, self._vectors, self._millis()]
 
        payload = msgpack.packb(frame, use_single_float=True, use_bin_type=True)
        size = len(payload)
        header = bytes([0xFD, size & 0xFF, (size >> 8) & 0xFF])
        
        out = header + payload
 
        if self._output is None:
            sys.stdout.buffer.write(out)
            sys.stdout.buffer.flush()
        else:
            self._output.write(out)
            
 
        self._reset()
 
    def is_empty(self) -> bool:
        return not self._has_data
 
    def _reset(self):
        self._sensors = []
        self._vectors = []
        self._has_data = False
 
    def _add_sensor_to_pack(self, s: Sensor):
        # [0]=name [1]=input [2]=isTriggered [3]=threshold [4]=layer [5]=x [6]=y
        self._sensors.append([
            s.name,
            s.input,
            s.is_triggered(),
            s.threshold,
            s.layer,
            s.x,
            s.y
        ])
        self._has_data = True
 
    def _add_vector_to_pack(self, v: Vector):
        # [0]=name [1]=rotation [2]=color [3]=layer [4]=x [5]=y
        self._vectors.append([
            v.name,
            v.rotation,
            v.color,
            v.layer,
            v.x,
            v.y
        ])
        self._has_data = True
 
    @staticmethod
    def _millis() -> int:
        return int((time.monotonic() - _START_TIME) * 1000)
 
